httpserver/http_connection.cpp

```diff
@@ -294,6 +300,43 @@ bool connection_t::verify_signature() {
     return ok;
 }
 
+void connection_t::process_message_test(uint64_t height,
+                                        const std::string& msg_hash) {
+
+    BOOST_LOG_TRIVIAL(debug)
+        << "Performing message test, attempt: " << repetition_count_;
+
+    std::string answer;
+
+    MessageTestStatus status =
+        service_node_.process_msg_test(height, msg_hash, answer);
+    if (status == MessageTestStatus::SUCCESS) {
+        delay_response_ = true;
+        body_stream_ << answer;
+        response_.result(http::status::ok);
+        this->write_response();
+    } else if (status == MessageTestStatus::RETRY) {
+        delay_response_ = true;
+        repetition_count_++;
+
+        repeat_timer_.expires_after(TEST_RETRY_PERIOD);
+        repeat_timer_.async_wait([self = shared_from_this(), height,
+                                  msg_hash](const error_code& ec) {
+            if (ec) {
+                if (ec != boost::asio::error::operation_aborted) {
+                    log_error(ec);
+                }
+            } else {
+                self->process_message_test(height, msg_hash);
+            }
+        });
+
+    } else {
+        response_.result(http::status::bad_request);
```

Comment:
I would only use `bad_request` if the request is invalid (e.g. could not be parsed)
and provide the test result in the body as json, with a 200 OK status.

Reply:
There is already a TODO to send a better error message. Not sure why `bad_request` is inappropriate here, we already send it when json is invalid for example, or the signature is bad. This case is similar: testee is incorrect which is the tester's fault.